niriPipe/utils/downloader.py

```diff
@@ -0,0 +1,333 @@
+        except Exception as e:
+            self.logger.error(
+                "Problem getting data urls; did the input table " +
+                "have a publisherID column?", exc_info=True
+                )
+            raise e
+
+        for url, pid in zip(urls, pids):
+            try:
+                filename = self._get_file(url)
+                self.logger.debug("Downloaded {}".format(filename))
+            except Exception as e:
+                self.logger.warning(
+                    "Frame {} failed to download.".format(pid),
+                    exc_info=True
+                )
+                continue
+        self._check_for_MD_files()
+
+    def _get_file(self, url):
+        """
+        Gets a file from the specified url and returns the filename.
+        """
+        r = requests.get(url, stream=True)
+        # Parse out filename from header
+        try:
+            filename = re.findall(
+                "filename=(.+)",
+                r.headers['Content-Disposition'])[0]
+            # Some filenames now have extra quotation marks
+            filename = filename.replace('"', '')
+        except KeyError:
+            # 'Content-Disposition' header wasn't found, so parse filename
+            # from URL. Typical URL looks like:
+            # https://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/data/pub/GEM/N20140505S0114.fits?RUNID=mf731ukqsipqpdgk
+            filename = (url.split('/')[-1]).split('?')[0]
+
+        # Write the fits file to the current directory, verifying the md5
+        # hash as we go. Store partial results in a temporary file.
+        self._write_with_temp_file(r, filename)
+
+        return filename
+
+    def _write_with_temp_file(self, response, filename):
+        """
+        Write the fits file, verifying the md5 hash as we go.
+        Store partial results in a temporary file.
+        """
+        try:
+            server_checksum = response.headers['Content-MD5']
+        except KeyError:
+            # Catch case that header didn't contain a 'content-md5' header
+            self.logger.warning(
+                "Content-MD5 header not found for file {}.".format(filename) +
+                " Skipping checksum validation."
+            )
+            server_checksum = None
+
+        # Write out content (first to a temp file) optionally doing md5 check.
+        download_checksum = hashlib.md5()
+        with tempfile.TemporaryFile(
+                mode='w+b',
+                prefix=filename,
+                dir=self.temp_downloads_path) as f:
+            for chunk in response.iter_content(chunk_size=128):
+                f.write(chunk)
+                download_checksum.update(chunk)
+            if server_checksum and \
+               (server_checksum != download_checksum.hexdigest()):
+                self.logger.error("Problem downloading {}.".format(filename))
+                raise IOError
+            f.seek(0)
+            with open(os.path.join(
```

Comment:
2 potential problems here:
1. you copy the file instead of just moving it which is faster
2. the temporary copy is on a different location. While it's convenient to use `temple` if source and destination are on mounts, even move could be a costly operation.

Solution is to save the tmp file in the same destination with a different name (prefix with "." for example) and change the name at the end. You need to take control of the cleanup so you'll probably have to surround the entire thing with a `try` and `finally` to make sure the temporary file is removed when function exits.